### ISSUE TITLE
IPAddress.GetHashCode allocates a string for IPv6

### ISSUE THREAD

**Issue by i3arnon** [dotnet/corefx#24117]
In order to return a hash code for IPv6 IPAddress calls (and caches) `ToString` and returns the result's hash code:
```csharp
public override int GetHashCode()
{
    if (IsIPv6)
    {
        if (_hashCode == 0)
        {
            _hashCode = StringComparer.OrdinalIgnoreCase.GetHashCode(ToString());
        }

        return _hashCode;
    }
    else
    {
        // ...
    }
}
```

These allocations constantly show up high on our list when we analyze our product's performance (even though the results are cached).
Is there any reason not to just calculate the hash code from the internal `ushort[] _numbers` and `_addressOrScopeId`?

A trivial implementation would be to simply xor all the values (but I can use something more complicated):

```csharp
public override int GetHashCode()
{
    if (IsIPv6)
    {
        if (_hashCode == 0)
        {
            int hashCode = (int)PrivateScopeId;
            for (int i = 0; i < NumberOfLabels; i += 2)
            {
                hashCode ^= _numbers[i] << 16 | _numbers[i + 1];
            }

            _hashCode = hashCode;
        }

        return _hashCode;
    }
    else
    {
        // ...
    }
}
```

I might be missing something obvious, but if not, I would love to remove these unnecessary allocations.

**Comment by davidsh** [comment]
>I might be missing something obvious, but if not, I would love to remove these unnecessary allocations.

This was done originally as a tradeoff of space vs. time.

**Comment by danmosemsft** [comment]
@i3arnon is this showing up high in elapsed time, or allocations?

> tradeoff of space vs. time.

`StringComparer.OrdinalIgnoreCase.GetHashCode` is managed code, seems a good hashcode could be generated at least as fast inline here. There is discussion [here](https://github.com/dotnet/corefx/issues/14354) over improving on xor (which is apparently not good)

**Comment by i3arnon** [comment]
> is this showing up high in elapsed time, or allocations?

Allocations, but I'm specifically looking for allocations, so I can't say much on elapsed time.

**Comment by danmosemsft** [comment]
@davidsh would you accept a PR if someone wanted to make a customized get hash code for IPAddresses?

**Comment by davidsh** [comment]
>would you accept a PR if someone wanted to make an (efficient) customized get hash code for IPAddresses?

Yes, assuming there are sufficient tests to validate it especially that it follows the generic rules of what GetHashCode() has to support.

We would also want to see perf results to understand how much "better" it is than we is currently implemented.

**Comment by danmosemsft** [comment]
@i3arnon are you interested in throwing up a PR, to trigger discussion, with some measurements? Benchmark.NEt seems generally used.

**Comment by i3arnon** [comment]
@danmosemsft I am. I'll start working on it in a couple of days.

**Comment by danmosemsft** [comment]
This is fixed by #24188
